doc/specs/#6899 - Action IDs/#6899 - Action IDs.md

```diff
@@ -0,0 +1,228 @@
+
+## Solution Design
+
+This spec was largely inspired by the following diagram from @DHowett:
+
+![figure 1](data-mockup.png)
+
+The goal is to introduce an `id` parameter by which actions could be uniquely
+refered to. If we'd ever like to use an action outside the list of `actions`, we
+can simply refer to the action's ID, allowing the user to only define the action
+_once_.
+
+We'll start by renaming `bindings` to `actions`. `bindings` was suggested as a
+rename for `keybindings` in [#6532], as a way to make the name more generic.
+Discussion with the team lead to the understanding that the name `actions` would
+be even better, as a way of making the meaning of the "list of actions" more
+obvious.
+
+When we're parsing `actions`, we'll make three passes:
+* The first pass will scan the list for objects with an `id` property. We'll
+  attempt to parse those entries into `ActionAndArgs` which we'll store in the
+  global `id->ActionAndArgs` map. If any entry doesn't have an `id` set, we'll
+  skip it in this phase. If an entry doesn't have a `command` set, we'll ignore
+  it in this pass.
+* The second pass will scan for _keybindings_. Any entries with `keys` set will
+  create a `KeyChord->ActionAndArgs` entry in the keybindings map. If the entry
+  has an `id` set, then we'll simply re-use the action we've already parsed for
+  the `id`, from the action map. If there isn't an `id`, then we'll parse the
+  action manually at this time. Entries without a `keys` set will be ignored in
+  this pass.
+* The final pass will be to generate _commands_. Similar to the keybindings
+  pass, we'll attempt to lookup actions for entries with an `id` set. If there
+  isn't an `id`, then we'll parse the action manually at this time. We'll then
+  get the name for the entry, either from the `name` property if it's set, or
+  the action's `GenerateName` method.
+
+For a visual representation, let's assume the user has the following in their
+`actions`:
+
+![figure 2](data-mockup-actions.png)
+
+We'll first parse the `actions` to generate the mapping of `id`->`Actions`:
+
+![figure 3](data-mockup-actions-and-ids.png)
+
+Then, we'll parse the `actions` to generate the mapping of keys to actions, with
+some actions already being defined in the map of `id`->`Actions`:
+
+![figure 4](data-mockup-actions-and-ids-and-keys.png)
+
+
+When layering `actions`, if a later settings file contains an action with the
+same `id`, it will replace the current value. In this way, users can redefine
+actions, or remove default ones (with something like `{ "id":
+"Terminal.OpenTab", "command":null }`
+
+We'd maintain a large list of default actions, each with unique `id`s set. These
+are all given `id`'s with a `Terminal.` prefix, to easily identify them as
+built-in, default actions. Not all of these actions will be given keys, but they
+will all be given `id`s.
+
+> 👉 NOTE: The IDs for the default actions will need to be manually created, not

```

Comment:
Just to clarify, when you say "manually created", do you mean we'll have to hard code it in the default settings?

Reply:
yes